components/Elevation/src/UIView+MaterialElevationResponding.m

```diff
@@ -0,0 +1,59 @@
+// Copyright 2019-present the Material Components for iOS authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "UIView+MaterialElevationResponding.h"
+
+#import "MDCElevation.h"
+
+@implementation UIView (MaterialElevationResponding)
+
+- (CGFloat)mdc_baseElevation {
+  CGFloat totalElevation = 0;
+  UIView *current = self;
+
+  while (current != nil) {
+    id<MDCElevation> elevatableCurrent = [current conformingObjectInResponderChain];
+    if (elevatableCurrent) {
+      if ([elevatableCurrent respondsToSelector:@selector(mdc_overrideBaseElevation)]) {
+        totalElevation += elevatableCurrent.mdc_overrideBaseElevation;
```

Comment:
Is it expected that simply *responding* to this API means that it must be used? What do you think if returning a negative value is equivalent to saying, "this is not set"?

Reply:
I believe having it optional it the means of having a `BOOL` that replaces _use this/don't use this_ if a client doesn't want it used then they shouldn't respond to that selector.

Reply:
In a subclass scenario clients can return the `mdc_currentElevation` for the `mdc_overrideElevation` property. So clients can always subclass and still get the same behavior if they want.

Reply:
Can you provide an example of how you think that would look? I've provided the only solution I can think of below. Please let me know if I've missed something.

```objc
- (CGFloat)mdc_overrideBaseElevation {
   return self.didSetOverrideBaseElevation ? [super mdc_overrideBaseElevation] : self.mdc_currentElevation;
}

- (void)setMDC_OverrideBaseElevation:(CGFloat)mdc_overrideBaseElevation {
  self.didSetOverrideBaseElevation = YES;
  [super setMDC_OverrideBaseElevation:mdc_overrideBaseElevation];
}
```

Reply:
```objc

@implementation MDCConformingMDCElevationOverrideViewSubclass

- (CGFloat)mdc_overrideBaseElevation {
  if (self.superview) {
    if ([self.superview conformsToProtocol:@protocol(MDCElevation)]) {
      UIView<MDCElevation> *superview = (UIView<MDCElevation> *)self.superview;
      return superview.mdc_currentElevation + superview.mdc_baseElevation +
             self.mdc_currentElevation;
    }
    return self.superview.mdc_baseElevation + self.mdc_currentElevation;
  }
  return 0;
}

@end

```